python/cudf/cudf/io/parquet.py

```diff
@@ -561,6 +567,116 @@ def read_parquet(
         **kwargs,
     )
 
+    # Apply filters row-wise (if any are defined), and return
+    return _apply_post_filters(df, filters)
+
+
+def _normalize_filters(filters: list | None) -> List[List[tuple]] | None:
+    # Utility to normalize and validate the `filters`
+    # argument to `read_parquet`
+    if filters is not None:
+        msg = (
+            f"filters must be None, or non-empty List[Tuple] "
+            f"or List[List[Tuple]]. Got {filters}"
+        )
+        if not filters or not isinstance(filters, list):
+            raise TypeError(msg)
+
+        def _validate_predicate(item):
+            if not isinstance(item, tuple) or len(item) != 3:
+                raise TypeError(
+                    f"Predicate must be Tuple[str, str, Any], "
+                    f"got {predicate}."
+                )
+
+        filters = filters if isinstance(filters[0], list) else [filters]
+        for conjunction in filters:
+            if not conjunction or not isinstance(conjunction, list):
+                raise TypeError(msg)
+            for predicate in conjunction:
+                _validate_predicate(predicate)
+
+    return filters
+
+
+def _handle_in(column, value, *, negate):
+    if not isinstance(value, (list, set, tuple)):
+        raise TypeError(
+            "Value of 'in' or 'not in' filter must be a list, set, or tuple."
+        )
+    if negate:
+        return reduce(operator.and_, (operator.ne(column, v) for v in value))
+    else:
+        return reduce(operator.or_, (operator.eq(column, v) for v in value))
```

Comment:
Ah, I hadn't really thought when writing this, let's do:

```suggestion
    return ~column.isin(value) if negate else column.isin(value)
```

To avoid scanning over the column once per value and then reducing.

Reply:
Regardless, the previous code should be spelled with `any`/`all` instead of reducing with or/and. Use builtin reductions when possible! And use `==` / `!=` instead of calling the (in)equality operator literally.

Reply:
As below, `any` and `all` want to produce a single boolean, this is producing a boolean mask column, so that can't be done (although yes, `column == v` is clearer).

Reply:
I modified things to use code like `column == v`. However, just a note that there is probably no performance advantage in doing this over the simpler `operator` approach. We still need to wrap the conditional in another function to reduce.

Reply:
Oh gosh, I apologize. I didn't realize these were acting on Series. 😪 What you have now is reasonable. I'll re-review this.

Reply:
I made two significant mistakes in my previous review comment, for which I am terribly sorry.

1. I was perfectly fine with the use of `operator.lt` in the handler dispatch. The only thing I wanted to replace was the reduction. I wasn't clear about that in my comment - the handler dispatch is an ideal use of `operator`.
2. My proposal of replacing `reduce(operator.or_, ...))` with `any(...)` is wrong. It was not a logical `or` that you intended, but _bitwise or_ `|` [which is the action of `operator.or_`](https://docs.python.org/3/library/operator.html#operator.or_). I mistakenly thought it was representing a logical or, which is the reduction operator of `any`.

I'd consider going back to the previous design here, since my suggestions made it worse. 🙃 It was partly a misreading of the code, and partly incorrect information about `operator` in my brain (which I didn't realize until digging deeper).

Reply:
>for which I am terribly sorry.

:)  No worries - I really appreciate having your eyes on this @bdice !

>I'd consider going back to the previous design here

Sure - I agree that operator makes for this case. I can roll that piece back.